### PR TITLE
[0.10.0] Elective Options In Student List Printout

### PR DESCRIPTION
**Pull request by CursorWind**
localhost_3000_en-US_classes_301_print(iPad Air)

## Features
- Added extra options in Student List Printout
  Resolves FRO-99 [(link)](https://linear.app/skiso/issue/FRO-99/elective-options-in-student-list-printout)
	- Added filtering options for students with allergies and students without elective subjects
     Resolves FRO-88
     - Hidden chosen elective column from students 
     Resolves FRO-86

### CODE REVIEW

File: components/classes/StudentListPrintout.tsx [hunk 1]
```diff
@@ -24,6 +24,8 @@ export type OptionsType = {
     | "shirtSize"
     | "pantsSize"
     | "elective"
+    | "noElective"
+    | "hasAllergies"
   )[];
```
Comment (SiravitPhokeed) [inline components/classes/StudentListPrintout.tsx:29]: I personally wouldn’t do this. Columns and filters are visually separated on the UI and so should they be in the code. It’s not wrong per se, just slightly confusing.

I would separate out `columns` and `filters`, like this:
```ts
export type OptionsType = {
  language: LangCode;
  columns: (
    | "classNo"
    | "studentID"
    | "prefix"
    | "fullName"
    | "nickname"
    | "allergies"
    | "shirtSize"
    | "pantsSize"
    | "elective"
  );
  filters: (
    | "noElective"
    | "hasAllergies"
  )[];
  numEmpty: number;
  enableNotes: boolean;
  enableTimestamp: boolean;
};
```

File: components/classes/StudentListTableBody.tsx [hunk 2]
```diff
@@ -4,6 +4,7 @@ import { Student } from "@/utils/types/person";
 import { list } from "radash";
 import { MAXIMUM_EMPTY_COLUMNS, OptionsType } from "@/components/classes/StudentListPrintout";
 import { StylableFC } from "@/utils/types/common";
+import permitted from "@/utils/helpers/permitted";
```
Comment (SiravitPhokeed) [inline components/classes/StudentListTableBody.tsx:7]: Where is this used?

File: components/classes/StudentsPrintOptions.tsx [hunk 3]
```diff
@@ -74,6 +76,27 @@ const StudentsPrintOptions: FC<{
...
         </FormGroup>
+
+        <FormGroup label={t("filterColumns.label")}>
+          {(
+            [
+              mysk.user &&
+                (mysk.user.is_admin || mysk.user.role !== UserRole.student) &&
+                "noElective",
+                "hasAllergies",
+            ].filter((column) => column) as OptionsType["columns"]
```
Comment (SiravitPhokeed) [inline components/classes/StudentsPrintOptions.tsx:87]: Here’s a clear reason why you should separate `filters` out of `columns`. If you were to put keys from `columns` into this array, TypeScript wouldn’t flag it as an error.

File: public/static/locales/en-US/classes.json [hunk 4]
```diff
@@ -54,6 +54,11 @@
       "pantsSize": "School pants size",
       "elective": "Chosen elective"
     },
+    "filterColumns": {
+      "label": "Filters",
+      "noElective": "Without elective",
+      "hasAllergies": "Allergies"
+    },
```
Comment (SiravitPhokeed) [inline public/static/locales/en-US/classes.json:61]: I’m not the biggest fan of this name change. I think `columns` is good as is, and `filterColumns` makes no sense. We are filtering rows, not columns, and I think `filters` on its own communicate enough. Try to be succinct.

File: components/classes/StudentListTableBody.tsx [hunk 5]
```diff
@@ -21,6 +22,8 @@ const StudentListTableBody: StylableFC<{
 }> = ({ studentList, options, style, className }) => (
   <tbody style={style} className={className}>
     {studentList.map((student) => (
+      !(options.columns.includes("noElective") && student.chosen_elective) && 
+      !(options.columns.includes("hasAllergies") && !student.allergies?.length) &&
```
Comment (SiravitPhokeed) [inline components/classes/StudentListTableBody.tsx:26]: This works, and I would even be OK with leaving it as is, but it’s not how I would do it if you’re curious. As I see it, there are many ways of approaching this issue:

### 1. Conditional rendering with short-circuit evaluation
You’ve gone with this, and I think it makes perfect sense. The downfall is that short-circuit evaluation indents the code, and the logic is a lot to wrap your head around, especially if you want to _hide_ an element when the condition matches.

### 2. Conditional rendering with an early return

This is easier to read as `return null` clearly communicates that you don’t want the rest of the code to execute. Here’s how this would look:

```tsx
{studentList.map((student) => {
  if (options.filters.includes("noElective") && student.chosen_elective)
    return null;
  // Rest of conditions...
  return (
    // Rest of code...
  );
})}
```

### 3. Filtering the source array

This is what I would have done. You can use [`filter()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) on the data source itself, `studentList`, before the array is mapped into table rows.

```tsx
{studentList
  .filter((student) => (
    (!options.filters.includes("noElective") || !student.chosen_elective) &&
    (!options.columns.includes("hasAllergies") || student.allergies?.length)
  ))
  .map((student) => (
    // Rest of code...
  ))
}
```

I have to admit that this is very similar to your code, which is why I think you’re approach is already good enough. In an attempt to make it a bit clearer, though, I used the principle of Implication Elimination (P ⇒ Q ≡ ¬P ∨ Q) instead.

TL;DR: This is good, but it can be done differently, if you wish.

**Review by SiravitPhokeed — CHANGES_REQUESTED** [review]
The preview works as advertised! I do want you to have a look at separating `columns` and `filters` in both `OptionsType` and the translation files, though. More details in comments.

File: components/classes/StudentListTableBody.tsx (hunk 2 above)
Comment (CursorWind) [reply components/classes/StudentListTableBody.tsx:7]: I originally thought this had to be used while editing, and didn't remove before the push. I'll remove that right away.

File: components/classes/StudentListPrintout.tsx (hunk 1 above)
Comment (CursorWind) [reply components/classes/StudentListPrintout.tsx:31]: Great suggestion for future usage and expansion, edited.

**Review by SiravitPhokeed — APPROVED** [review]
You didn’t run Prettier on some files you worked on. **Don’t forget to install Prettier and run it on all files you’ve touched before submitting a PR.**

I’ve already run Prettier for you, so you don’t have to do anything, for now. LGTM!
